synapse/util/iterutils.py

```diff
@@ -46,3 +48,39 @@ def chunk_seq(iseq: ISeq, maxlen: int) -> Iterable[ISeq]:
     If the input is empty, no chunks are returned.
     """
     return (iseq[i : i + maxlen] for i in range(0, len(iseq), maxlen))
+
+
+def sorted_topologically(
+    nodes: Iterable[T], graph: Dict[T, Collection[T]],
+) -> Generator[T, None, None]:
+    """Given a set of nodes and a graph, yield the nodes in toplogical order.
+
+    For example `sorted_topologically([1, 2], {1: [2]})` will yield `2, 1`.
+    """
+
+    # This is implemented by Kahn's algorithm.
+
+    degree_map = {node: 0 for node in nodes}
+    reverse_graph = {}  # type: Dict[T, Set[T]]
+
+    for node, edges in graph.items():
+        if node not in degree_map:
+            continue
+
+        for edge in edges:
+            if edge in degree_map:
+                degree_map[node] += 1
+
+            reverse_graph.setdefault(edge, set()).add(node)
+        reverse_graph.setdefault(node, set())
+
+    zero_degree = [node for node, degree in degree_map.items() if degree == 0]
```

Comment:
This might not return things in the same order depending on equivalent (?) inputs:

```pycon
>>> list(sorted_topologically([2, 1], {1: [], 2: []}))
[1, 2]
>>> list(sorted_topologically([1, 2], {1: [], 2: []}))
[2, 1]
```

I'm not sure if these are meant to be stable inputs or not though.

It'd probably be good to see some tests for `sorted_topologically` while we're here!

Reply:
Yeah, topological orderings aren't unique, and we don't really care in this case. I'll have a quick look to see if its easy to make it stable at least, though it'll probably always depend on the ordering of the inputs

Reply:
I suppose as long as it is stable depending on the inputs it should be OK.